### PR TITLE
fix: use BNs for swaps balance rather than .toNumber conversion

### PR DESCRIPTION
**Pull request by svvald**

### CODE REVIEW

File: src/components/common/Swap/Swap.tsx [hunk 1]
```diff
@@ -98,10 +98,10 @@ const Swap = () => {
     }
   }, [isConnected]);
 
-  const sellBalance = balances?.find((b) => b.assetId === coinsConfig.get(swapState.sell.coin)?.assetId)?.amount.toNumber();
-  const sellBalanceValue = sellBalance ? sellBalance / 10 ** coinsConfig.get(swapState.sell.coin)?.decimals! : 0;
-  const buyBalance = balances?.find((b) => b.assetId === coinsConfig.get(swapState.buy.coin)?.assetId)?.amount.toNumber();
-  const buyBalanceValue = buyBalance ? buyBalance / 10 ** coinsConfig.get(swapState.buy.coin)?.decimals! : 0;
+  const sellBalance = balances?.find((b) => b.assetId === coinsConfig.get(swapState.sell.coin)?.assetId)?.amount;
+  const sellBalanceValue = sellBalance ?? new BN(0);
```
Comment (mpoplavkov) [inline src/components/common/Swap/Swap.tsx:102]: Is it intentional to remove the decimals accounting here and below?

File: src/components/common/Swap/Swap.tsx [hunk 2]
```diff
@@ -345,7 +345,7 @@ const Swap = () => {
     refetchBalances
   ]);
 
-  const insufficientSellBalance = parseFloat(sellValue) > sellBalanceValue;
+  const insufficientSellBalance = sellBalanceValue.lt(parseFloat(sellValue));
```
Comment (mpoplavkov) [inline src/components/common/Swap/Swap.tsx:348]: So `sellBalanceValue` is a decimal now and `sellValue` is float? Should it be like this?

File: src/components/common/Swap/components/CurrencyBox/CurrencyBox.tsx [hunk 3]
```diff
@@ -45,19 +47,19 @@ const CurrencyBox = ({ value, coin, mode, balance, setAmount, loading, onCoinSel
   const handleMaxClick = useCallback(() => {
     let amountStringToSet;
     if (coin === "ETH" && mode === "sell") {
-      const amountWithoutGasFee = balance - MinEthValue;
-      amountStringToSet = amountWithoutGasFee > 0 ? amountWithoutGasFee.toFixed(9) : balance.toString();
+      const amountWithoutGasFee = balance.sub(MinEthValueBN);
+      amountStringToSet = amountWithoutGasFee.gt(0)
+        ? amountWithoutGasFee.formatUnits(decimals)
```
Comment (mpoplavkov) [inline src/components/common/Swap/components/CurrencyBox/CurrencyBox.tsx:52]: Is it the same as `format({ units:`? Should we keep just one style?

File: src/components/common/Swap/Swap.tsx (hunk 1 above)
Comment (mpoplavkov) [reply src/components/common/Swap/Swap.tsx:102]: Found that we account for this in currency box

File: src/components/common/Swap/Swap.tsx (hunk 2 above)
Comment (svvald) [reply src/components/common/Swap/Swap.tsx]: Fixed

File: src/components/common/Swap/components/CurrencyBox/CurrencyBox.tsx (hunk 3 above)
Comment (svvald) [reply src/components/common/Swap/components/CurrencyBox/CurrencyBox.tsx:52]: Fixed

File: src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx [hunk 4]
```diff
@@ -54,15 +53,13 @@ const CoinsListModal = ({ selectCoin, balances }: Props) => {
...
-
-        if (bBalanceValue !== aBalanceValue) {
-          return bBalanceValue - aBalanceValue;
+        const aBalance = balances.find((b) => b.assetId === coinA.assetId)?.amount ?? new BN(0);
+        const bBalance = balances.find((b) => b.assetId === coinB.assetId)?.amount ?? new BN(0);
+
+        const balancesDifference = aBalance.sub(bBalance);
+
+        if (!balancesDifference.eq(0)) {
+          return balancesDifference.gt(0) ? -1 : 1;
```
Comment (mpoplavkov) [inline src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx:62]: nit: simplification
```ts
        if (!aBalance.eq(bBalance)) {
          return aBalance.gt(bBalance) ? -1 : 1;
        }
```

File: src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx [hunk 5]
```diff
@@ -54,15 +53,13 @@ const CoinsListModal = ({ selectCoin, balances }: Props) => {
...
-        const aBalance = balances.find((b) => b.assetId === coinA.assetId)?.amount.toNumber();
-        const aBalanceValue = aBalance ? aBalance / 10 ** aDecimals : 0;
-        const bDecimals = coinsConfig.get(coinB.name)?.decimals!;
-        const bBalance = balances.find((b) => b.assetId === coinB.assetId)?.amount.toNumber();
-        const bBalanceValue = bBalance ? bBalance / 10 ** bDecimals : 0;
-
-        if (bBalanceValue !== aBalanceValue) {
-          return bBalanceValue - aBalanceValue;
+        const aBalance = balances.find((b) => b.assetId === coinA.assetId)?.amount ?? new BN(0);
+        const bBalance = balances.find((b) => b.assetId === coinB.assetId)?.amount ?? new BN(0);
```
Comment (mpoplavkov) [inline src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx:57]: We need to account for decimals when doing the comparison

**Review by mpoplavkov — CHANGES_REQUESTED** [review]

File: src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx (hunk 4 above)
Comment (svvald) [reply src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx]: Fixed

File: src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx (hunk 5 above)
Comment (svvald) [reply src/components/common/Swap/components/CoinsListModal/CoinsListModal.tsx]: Fixed

**Review by mpoplavkov — APPROVED** [review]

**Pull request merged**